### PR TITLE
implement C ABI lowering for CSKY

### LINKED ISSUE #551: wrong define of 128-bit related functions like `__multi3` and `__floatuntidf` on `csky`
I have added `csky-unknown-linux-gnuabiv2` and `csky-unknown-linux-gnuabiv2hf` last days in https://github.com/rust-lang/rust/pull/113658 . And I got `SIGSEGV` when I run a executable file on csky.

I checked the file and found that in `LLVM-IR` file of `compiler-buildins`, (Taking `__floatuntidf` as an example),it defines `__floatuntidf`  as 
```
define hidden noundef double @__floatuntidf(ptr noalias nocapture noundef readonly align 4 dereferenceable(16) %i) unnamed addr 
```
which has a `ptr` parameter but not values.

In executable files, the value of parameters are passed to the `__floatuntidf`. and then it regard the value as an address to load data, which causes the `SIGSEGV`

---

It's confused me that what caused the parameters become `ptr`. 
Is this function compiled from [float/conv.rs](https://github.com/rust-lang/compiler-builtins/blob/d1d785a538d51d0082c3e6209a28c9fb1a90ec3c/src/float/conv.rs#L99C1-L104) or anywhere else?
And how can I fixed it?

### PR DESCRIPTION
**Pull request by Dirreke**
fix https://github.com/rust-lang/compiler-builtins/issues/551

 
 ​Reference: [CSKY ABI Manual](https://occ-oss-prod.oss-cn-hangzhou.aliyuncs.com/resource//1695027452256/T-HEAD_800_Series_ABI_Standards_Manual.pdf)
 ​
Reference: [Clang CSKY lowering code](https://github.com/llvm/llvm-project/blob/4a074f32a6914f2a8d7215d78758c24942dddc3d/clang/lib/CodeGen/Targets/CSKY.cpp#L76-L162)

r? @bjorn3

### CODE REVIEW

File: compiler/rustc_target/src/abi/call/csky.rs [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+fn classify_ret<Ty>(arg: &mut ArgAbi<'_, Ty>) {
+    // For argument type, the first 4*XLen parts of aggregate will be passed
+    // in registers, and the rest will be passed in stack.
+    // So we can coerce to integers directly and let backend handle it correctly.
+    // For return type, aggregate which <= 2*XLen will be returned in registers.
+    // Otherwise, aggregate will be returned indirectly.
+    if arg.layout.is_aggregate() {
+        let total = arg.layout.size;
+        if total.bits() > 64 {
+            // We rely on the LLVM backend lowering code to lower passing a scalar larger than 2*XLEN.
```
Comment (bjorn3) [inline compiler/rustc_target/src/abi/call/csky.rs:18]: I think this comment is misplaced.

File: compiler/rustc_target/src/abi/call/csky.rs [hunk 2]
```diff
@@ -1,17 +1,40 @@
...
 
-fn classify_ret<Ty>(ret: &mut ArgAbi<'_, Ty>) {
-    if ret.layout.is_aggregate() || ret.layout.size.bits() > 64 {
-        ret.make_indirect();
+use crate::abi::call::{ArgAbi, FnAbi, Reg, Uniform};
+
+fn classify_ret<Ty>(arg: &mut ArgAbi<'_, Ty>) {
+    // For argument type, the first 4*XLen parts of aggregate will be passed
+    // in registers, and the rest will be passed in stack.
+    // So we can coerce to integers directly and let backend handle it correctly.
```
Comment (bjorn3) [inline compiler/rustc_target/src/abi/call/csky.rs:12]: Same for this.

**Comment by bjorn3** [comment]
The Clang abi code is a bit hard to follow, but I think you correctly translated it. You could try [abi-cafe](https://github.com/gankra/abi-cafe) to see if there are mismatches. Abi-cafe is not exhaustive, but it should catch a fair amount of abi issues.

LGTM with the comments fixed.

**Comment by Dirreke** [comment]
I tried `abi-cafe`. However, it seems that it cannot use for cross-compoile now.

**Comment by bjorn3** [comment]
Right, it uses dlopen of the dylibs to test. Roughly how fast are CSKY processors? Is the perforance more like that of a microcontroller where you would never be able to run rustc on it, is it more like an SBC like a raspberry pi where you could cross-compile rustc for CSKY and then run abi-cafe using said CSKY rustc albeit rather slowly or does it have a fairly decent performance?

**Comment by Dirreke** [comment]
It have a decent performance. 

However, there's no gcc host tool on it. The open-source code of gcc is very old too. I will ask the maintainer of the csky if they can give the newest source code or gcc host.

(Besides, thanks for your guidance during this period

**Comment by Dirreke** [comment]
I cannot build rustc for csky too. `csky-linux-x86_64-gcc-toolchain` is based on `gcc 6.3.0` and llvm requires that gcc version must be at least 7.4.

Therefore, I won't test the abi with `abi-cafe` for now. This is enough to compile and run a sufficiently complex program with almost all hardware drivers related operations, mathematical optimization operations and web server.

Thanks all the time!
